### ISSUE TITLE
[ONNX] Support torch.bucketize

### ISSUE THREAD

**Issue by fairydora** [pytorch/pytorch#85935]
### 🚀 The feature, motivation and pitch

Hello,
Are there any plans for onnx to support torch.bucketize?

**Comment by justinchuby** [comment]
Curious where it is used / what your use case is?

**Comment by fairydora** [comment]
I need to resample my signal - interpolate features from having a shorter time sequence to match a longer sequence. The shorter one comes out of the audio features extraction (HuBERT), where the sequence's time dimension has shrunk because the transformer model is using a stack of convnets. I need to do an interpolation to get back to the fps we're using when capturing 4D meshes. I've previously used np.interp but torch script doesn't support numpy. Tried to use torchaudio.resample too, but even if it was to successfully compile (which it doesn't), the results will not be accurate for small frame rates due to the required rounding of the input fps to int.  I need a torch implementation and bucketize is a part of it.

**Comment by justinchuby** [comment]
Does `torch.nn.functional.interpolate` work for you?

**Comment by fairydora** [comment]
Hi Justin,
Indeed, it sort of hasn't worked for me, but it's possible I'm doing something wrong.
I previously dismissed this solution because of 2 reasons:
1. Less important reason.
The algorithm is different than the one I have used in an already trained, ready for deployment model. It gives me slightly different interpolation results - using it would require me to retrain the model (10 days) instead of simply exporting to onnx.

2. Crucial reason.
The issue I've had with torch.nn.functional.interpolate is that it seems to be coding my temporal dimension in a static way. 
I'm defining a dynamic temporal axis, and I've put the decorator @torch.jit.script above my interpolation function, hoping to compile all the functions called from that function with Torch Script. 

**The time dimension of the output consistently comes back with the same length as the original dummy input that was used in tracing.** 
I've written some simplified code to illustrate this (below). Please let me know if something jumps out at you immediately, I may be doing something wrong and wrongly dismissing a good solution to my problem. Thank you for your help and advice :) 



import onnxruntime
import torch
import numpy as np

# _____________________________________
# THIS IS MY INTERPOLATION FUNCTION, USING torch.nn.functional.interpolate
@torch.jit.script
def interpolate_features_torch_fnct(features: torch.Tensor, audio_time: float, output_fps: float = 60):
    n_features = features.shape[-1]
    output_n_samples = int(audio_time * output_fps)

    # features tensor: I need to permute to be able to smooth the features across the temporal dimension
    before = features.permute(0, 1, 3, 2)
    after = torch.nn.functional.interpolate(before, size=(n_features, output_n_samples), mode='bilinear')
    output = after.permute(0, 1, 3, 2)
    return output

# _____________________________________
# DEFINING A VERY BASIC MODEL, JUST FOR THE TEST.
class DummyModel(torch.nn.Module):
    def __init__(self):
        super(DummyModel, self).__init__()

    def forward(self, input):
        """
        I need to resample the output at a lower fps to the input of another module which is 60 fps.
        """
        fps = 40.
        output_fps = 60.
        n_frames = input.shape[2]
        audio_time = float(n_frames * (1/fps))
        output = interpolate_features_torch_fnct(features=input, audio_time=audio_time, output_fps=output_fps)
        return output

model = DummyModel()

# _____________________________________
# EXPORTING THE MODEL TO ONNX:
input_shape = (1, 1, 30, 4) #<- required 4D input for mode bilinear
dummy_input = torch.randn(*input_shape)
onnx_path = "onnx_test_interp.onnx"
# defining the temporal axis to be dynamic - the input sequence will have a different length each time
dynamic_axes = {'input': {2: 'seq_len'},
                'output': {2: 'seq_len'}} #<- I've also tested with ALL dims set as dynamic axes
torch.onnx.export(model,
                  dummy_input,
                  onnx_path,
                  verbose=False,
                  input_names=["input"],
                  output_names=["output"],
                  export_params=True,
                  dynamic_axes=dynamic_axes,
                  opset_version=13
                  )
# _____________________________________
# CHECK IF THE RESULTS FROM ONNX MATCH THE RESULTS FROM THE MODEL:
def compare_outputs(dummy_input_t, audio_time=1):
    # - output from the onnx model
    ort_session = onnxruntime.InferenceSession(onnx_path)
    dummy_input = dummy_input_t.detach().float().cpu().numpy()
    input_name = ort_session.get_inputs()[0].name
    onnx_inputs = {input_name: dummy_input}
    outputs = ort_session.run(None, onnx_inputs)
    ort_out = outputs[0]
    output_model = model(dummy_input_t)
    model_out = output_model.detach().cpu().numpy()
    return model_out, ort_out


#------- check the same length as was used during tracing.
input_shape = (1, 1, 30, 4)
print ("checking shape ", input_shape)
dummy_input = torch.randn(*input_shape)

model_out, ort_out = compare_outputs(dummy_input)
print("model_out.shape = ", model_out.shape)
print("ort_out.shape = ", ort_out.shape)
assert (model_out.shape == ort_out.shape)
np.testing.assert_allclose(model_out, ort_out, rtol=0.001, atol=0.001)
print("-> all good.")

#-------  check different sequence length.
input_shape = (1, 1, 150, 4)
print ("checking shape ", input_shape)
dummy_input = torch.randn(*input_shape)

model_out, ort_out = compare_outputs(dummy_input)
print("model_out.shape = ", model_out.shape)
print("ort_out.shape = ", ort_out.shape)
assert (model_out.shape == ort_out.shape)
np.testing.assert_allclose(model_out, ort_out, rtol=0.001, atol=0.001)
print("-> all good.")

"""
OUTPUTS:
checking shape  (1, 1, 30, 4) #<- this is the shape of the dummy input used at export
model_out.shape =  (1, 1, 45, 4)
ort_out.shape =  (1, 1, 45, 4)
-> all good.

checking shape  (1, 1, 150, 4) #<- using a different length of the input than what I used at export
model_out.shape =  (1, 1, 225, 4)
ort_out.shape =  (1, 1, 45, 4)       #<-  this output was supposed to be (1, 1, 225, 4)!
assert (model_out.shape == ort_out.shape)
AssertionError
"""

**Comment by justinchuby** [comment]
I made a few changes to your script:

1. `torch.jit.scripting` looks like is complaining about the input type to interpolate. `interpolate` supports Tensors as inputs to the size parameter but it is only declared to accept `int`s. I used `tracing` instead for this part of the model.
2. You want to take the shape using `Tensor.size()` instead of `shape[]`, because the former will give you a tensor when tracing, which is what makes the resulting shape dynamic.
3. There is a few explicit type conversion I had to make (`float` etc.) because torch.onnx did not realize implicit type conversion was done in the tensors @BowenBao . I recommend that you make these conversion explicit all the time.

LMK if you have any questions!

```diff
@@ -2,16 +2,20 @@
 import torch
 import numpy as np

 # _____________________________________
 # THIS IS MY INTERPOLATION FUNCTION, USING torch.nn.functional.interpolate
-@torch.jit.script
-def interpolate_features_torch_fnct(features: torch.Tensor, audio_time: float, output_fps: float = 60):
-    n_features = features.shape[-1]
-    output_n_samples = int(audio_time * output_fps)
+def interpolate_features_torch_fnct(features: torch.Tensor, audio_time: torch.Tensor, output_fps: float = 60):
+    n_features = features.size(-1)
+    if torch.jit.is_tracing():
+        output_n_samples = torch.floor(audio_time * output_fps).long()
+    else:
+        output_n_samples = int(audio_time * output_fps)

     # features tensor: I need to permute to be able to smooth the features across the temporal dimension
     before = features.permute(0, 1, 3, 2)
-    after = torch.nn.functional.interpolate(before, size=(n_features, output_n_samples), mode='bilinear')
+    after = torch.nn.functional.interpolate(before, size=[n_features, output_n_samples], mode='bilinear')
     output = after.permute(0, 1, 3, 2)
     return output

@@ -27,8 +31,13 @@
         """
         fps = 40.
         output_fps = 60.
-        n_frames = input.shape[2]
-        audio_time = float(n_frames * (1/fps))
+        n_frames = input.size(2)
+        if torch.jit.is_tracing():
+            # Dirty hack for tracing because
+            # n_frames is a tensor when tracing, an int when in eager mode
+            audio_time = n_frames.float() * (1/fps)
+        else:
+            audio_time = n_frames * (1/fps)
         output = interpolate_features_torch_fnct(features=input, audio_time=audio_time, output_fps=output_fps)
         return output

@@ -52,6 +61,7 @@
                   dynamic_axes=dynamic_axes,
                   opset_version=13
                   )
 # _____________________________________
 # CHECKING IF THE RESULTS FROM ONNX MATCH THE RESULTS FROM THE MODEL:
 def compare_outputs(dummy_input_t, audio_time=1):
```

```python
checking shape  (1, 1, 30, 4)
model_out.shape =  (1, 1, 45, 4)
ort_out.shape =  (1, 1, 45, 4)
-> all good.
checking shape  (1, 1, 150, 4)
model_out.shape =  (1, 1, 225, 4)
ort_out.shape =  (1, 1, 225, 4)
-> all good.
```

<img width="1060" alt="image" src="https://user-images.githubusercontent.com/11205048/193479737-fab089b0-680c-48e4-896f-5e8cce83a509.png">

**Comment by fairydora** [comment]
Hello Justin,
Thank you very much for taking the time to make these corrections, this solution is indeed working now :) This is a huge help, and I really appreciate how fast you've replied! 
Have a nice day :)

**Comment by AllenTiTaiWang** [comment]
Closing as the issue has been solved
